File: ui/narrador.py
"""Narração por voz do corpo focado, usando o sintetizador do sistema.

Roda em **thread separada**: ``pyttsx3.say()`` + ``runAndWait()`` bloqueiam até a
fala terminar, e no loop de render isso congelaria a cena por segundos — o mesmo
motivo pelo qual a captura de vídeo já roda fora do laço principal.

A fila guarda no máximo um pedido: trocando de planeta rápido, o mais recente
substitui o anterior em vez de enfileirar uma narração que ninguém quer ouvir.

Sem o pyttsx3 instalado (ou sem voz no sistema) o narrador simplesmente fica
inativo — nada quebra, o aplicativo segue mudo.
"""

from __future__ import annotations

import io
import queue
import threading
from decimal import ROUND_HALF_UP, Decimal

import pygame

from config import (
    ELEVENLABS_VOZ_NOME,
    IDIOMA_NARRACAO,
    NARRACAO_ATIVA_PADRAO,
    NARRAR_FICHA_COMPLETA,
    VELOCIDADE_NARRACAO,
    VOLUME_NARRACAO,
)
from dados.planetas import CorpoCeleste
from ui.voz_elevenlabs import SinteseElevenLabs

try:  # pragma: no cover - depende do ambiente
    import pyttsx3

    PYTTSX3_DISPONIVEL = True
    ERRO_IMPORT_PYTTSX3 = ""
except ImportError as _erro:  # pragma: no cover
    pyttsx3 = None  # type: ignore[assignment]
    PYTTSX3_DISPONIVEL = False
    ERRO_IMPORT_PYTTSX3 = f"{type(_erro).__name__}: {_erro}"

# Sentinela que encerra a thread.
_PARAR = object()


# Artigo definido antes do nome. Em português corrente dizemos "o Sol" e "a
# Terra", mas "Marte" e "Júpiter" dispensam artigo.
ARTIGO_DEFINIDO: dict[str, str] = {"Sol": "O", "Terra": "A", "Lua": "A"}

# Descrição do tipo já com artigo indefinido, para a frase fechar concordância.
TIPO_NARRADO: dict[str, str] = {
    "estrela": "uma estrela",
    "rochoso": "um planeta rochoso",
    "gasoso": "um gigante gasoso",
    "satelite": "um satélite natural",
}


def _numero(valor: float, casas: int = 0) -> str:
    """Número no padrão pt-BR, que é como o sintetizador lê corretamente.

    Arredonda com ROUND_HALF_UP para bater com o ``Intl.NumberFormat`` da versão
    web. O padrão do Python é half-even ("banker's rounding"), que em empates
    exatos discorda do JavaScript: a rotação de Mercúrio (58,65 dias) virava
    58,6 aqui e 58,7 lá — o verificador de paridade pegou a diferença.
    """
    arredondado = Decimal(str(valor)).quantize(
        Decimal(1) if casas == 0 else Decimal("0." + "0" * casas),
        rounding=ROUND_HALF_UP,
    )
    texto = f"{arredondado:,.{casas}f}"
    return texto.replace(",", "@").replace(".", ",").replace("@", ".")


def frases_da_ficha(corpo: CorpoCeleste) -> list[str]:
    """A ficha do corpo em orações completas, prontas para serem lidas.

    É a mesma informação do card, dita em português corrente: "Tem 12.756
    quilômetros de diâmetro" em vez de "Diâmetro equatorial: 12.756 km". Rótulo
    e valor soltos soariam como uma planilha sendo lida em voz alta.
    """
    frases = [f"Tem {_numero(corpo.diametro_km)} quilômetros de diâmetro."]

    if corpo.eh_satelite:
        frases.append(
            f"Fica a {_numero(corpo.distancia_km)} quilômetros "
            f"{'da' if corpo.orbita_em_torno_de == 'Terra' else 'de'} "
            f"{corpo.orbita_em_torno_de}."
        )
    elif corpo.distancia_ua > 0:
        frases.append(
            f"Fica a {_numero(corpo.distancia_ua, 2)} unidades astronômicas do Sol, "
            f"ou seja, {_numero(corpo.distancia_km)} quilômetros."
        )

    if corpo.periodo_orbital_dias > 0:
        if corpo.periodo_orbital_dias >= 365.26:
            anos = corpo.periodo_orbital_dias / 365.26
            frases.append(f"Uma volta completa leva {_numero(anos, 1)} anos terrestres.")
        else:
            frases.append(
                f"Uma volta completa leva {_numero(corpo.periodo_orbital_dias)} dias."
            )

    horas = abs(corpo.periodo_rotacao_horas)
    sentido = " no sentido contrário ao dos demais" if corpo.periodo_rotacao_horas < 0 else ""
    if horas >= 48:
        frases.append(f"Gira em torno de si mesmo em {_numero(horas / 24, 1)} dias{sentido}.")
    else:
        frases.append(f"Gira em torno de si mesmo em {_numero(horas, 1)} horas{sentido}.")

    if corpo.luas == 1:
        frases.append("Tem uma lua conhecida.")
    elif corpo.luas > 1:
        frases.append(f"Tem {_numero(corpo.luas)} luas conhecidas.")

    frases.append(
        f"A temperatura média é de {_numero(corpo.temperatura_media_c)} graus Celsius."
    )
    frases.append(corpo.fato_curioso)
    return frases


def texto_do_corpo(corpo: CorpoCeleste) -> str:
    """Texto narrado ao focar um corpo.

    A abertura é uma **oração completa** ("A Terra é um planeta rochoso.") e não
    uma lista de termos ("Terra. Planeta rochoso"). O motivo é prático: o modelo
    de voz identifica o idioma pelo texto, e nomes latinos soltos são ambíguos —
    medindo com a transcrição da própria ElevenLabs, "Sol. Estrela" saía em
    espanhol e "Marte. Planeta rochoso" em inglês. O verbo "é" acentuado e os
    artigos são o que ancora a frase no português.
    """
    artigo = ARTIGO_DEFINIDO.get(corpo.nome)
    sujeito = f"{artigo} {corpo.nome}" if artigo else corpo.nome
    descricao = TIPO_NARRADO.get(corpo.tipo, "um corpo celeste")
    partes = [f"{sujeito} é {descricao}."]
    if NARRAR_FICHA_COMPLETA:
        partes.extend(frases_da_ficha(corpo))
    return " ".join(partes)


class Narrador:
    """Fala frases curtas sem bloquear o loop de render."""

    def __init__(self, ativo: bool = NARRACAO_ATIVA_PADRAO) -> None:
        # Voz neural primeiro, voz do sistema como rede de proteção: qualquer
        # uma das duas já torna a narração possível.
        self._neural = SinteseElevenLabs()
        self.disponivel = PYTTSX3_DISPONIVEL or self._neural.disponivel
        self.ativo = ativo and self.disponivel
        self.mensagem = "" if self.disponivel else "TTS indisponível"
        self._fila: queue.Queue = queue.Queue(maxsize=1)
        self._thread: threading.Thread | None = None
        self._encerrando = threading.Event()

    # ------------------------------------------------------------- ciclo
    def iniciar(self) -> None:
        """Sobe a thread de fala (não faz nada se o TTS não existe)."""
        # Qual voz está em uso é a primeira dúvida quando a narração "não
        # funciona": o log evita ter que adivinhar.
        if self._neural.disponivel:
            print(
                f"[narrador] voz neural {ELEVENLABS_VOZ_NOME} (ElevenLabs) ativa",
                flush=True,
            )
        else:
            print(f"[narrador] ElevenLabs fora: {self._neural.mensagem}", flush=True)
            print(
                f"[narrador] voz local do sistema "
                f"{'disponível' if PYTTSX3_DISPONIVEL else 'INDISPONÍVEL'}"
                f"{'' if PYTTSX3_DISPONIVEL else f' ({ERRO_IMPORT_PYTTSX3})'}",
                flush=True,
            )
        if not self.disponivel or self._thread is not None:
            return
        self._thread = threading.Thread(
            target=self._laco, name="narrador", daemon=True
        )
        self._thread.start()

    def parar(self) -> None:
        """Encerra a thread, aguardando a fala em andamento terminar."""
        self._encerrando.set()
        if self._thread is None:
            return
        self._descartar_pendente()
        try:
            self._fila.put_nowait(_PARAR)
        except queue.Full:
            pass
        self._thread.join(timeout=3.0)
        self._thread = None

    def alternar(self) -> bool:
        """Liga/desliga a narração e devolve o novo estado."""
        if not self.disponivel:
            return False
        self.ativo = not self.ativo
        if not self.ativo:
            self._descartar_pendente()
        return self.ativo

    # ------------------------------------------------------------ pedidos
    def anunciar(self, texto: str) -> None:
        """Coloca uma frase na fila, substituindo a que estiver esperando."""
        if not self.ativo or self._thread is None:
            return
        self._descartar_pendente()
        try:
            self._fila.put_nowait(texto)
        except queue.Full:  # corrida com a thread: a fala atual já basta
            pass

    def _descartar_pendente(self) -> None:
        """Esvazia a fila para o pedido novo não entrar atrás de um velho."""
        while True:
            try:
                self._fila.get_nowait()
            except queue.Empty:
                return

    # --------------------------------------------------------------- thread
    def _laco(self) -> None:
        """Consome a fila falando uma frase por vez.

        Dois caminhos: a voz neural da ElevenLabs (melhor resultado, precisa de
        chave e rede) e a voz local do sistema. O segundo é o que garante que o
        aplicativo continue completo offline.
        """
        motor = self._criar_motor()
        if motor is None and not self._neural.disponivel:
            self.ativo = False
            self.disponivel = False
            return
        try:
            while not self._encerrando.is_set():
                try:
                    item = self._fila.get(timeout=0.2)
                except queue.Empty:
                    continue
                if item is _PARAR:
                    return
                if self._falar_neural(item):
                    continue
                if motor is None:
                    continue
                try:
                    motor.say(item)
                    motor.runAndWait()
                except RuntimeError:
                    # runAndWait reentrante ou motor derrubado pelo driver:
                    # perder uma fala é melhor que derrubar a thread.
                    continue
        finally:
            if motor is not None:
                try:
                    motor.stop()
                except Exception:  # noqa: BLE001 - encerramento não pode falhar
                    pass

    def _falar_neural(self, texto: str) -> bool:
        """Toca a frase com a voz da ElevenLabs. False = usar a voz local."""
        if not self._neural.disponivel:
            return False
        dados = self._neural.audio(texto)
        if not dados:
            return False
        try:
            if not pygame.mixer.get_init():
                pygame.mixer.init()
            # music (e não Sound) porque o retorno é MP3 e precisamos poder
            # cortar a fala anterior ao trocar de planeta.
            pygame.mixer.music.load(io.BytesIO(dados), "mp3")
            pygame.mixer.music.set_volume(VOLUME_NARRACAO)
            pygame.mixer.music.play()
        except pygame.error as erro:
            print(f"[narrador] falha ao tocar o áudio: {erro}", flush=True)
            return False

        # Espera a fala terminar, mas acordando com frequência: um pedido novo
        # na fila deve interromper a narração atual, não entrar atrás dela.
        while pygame.mixer.music.get_busy() and not self._encerrando.is_set():
            if not self._fila.empty():
                pygame.mixer.music.stop()
                break
            pygame.time.wait(50)
        return True

    def _criar_motor(self):
        """Instancia o pyttsx3 e escolhe a voz do idioma configurado."""
        if not PYTTSX3_DISPONIVEL:
            return None
        try:
            motor = pyttsx3.init()
        except Exception as erro:  # noqa: BLE001 - driver de TTS varia por SO
            self.mensagem = f"TTS indisponível: {erro}"
            print(f"[narrador] {self.mensagem}", flush=True)
            return None

        motor.setProperty("rate", VELOCIDADE_NARRACAO)
        motor.setProperty("volume", VOLUME_NARRACAO)
        voz = self._escolher_voz(motor)
        if voz is not None:
            motor.setProperty("voice", voz)
        else:
            print(
                f"[narrador] sem voz {IDIOMA_NARRACAO} instalada — usando a padrão",
                flush=True,
            )
        return motor

    @staticmethod
    def _escolher_voz(motor) -> str | None:
        """Procura uma voz do idioma configurado entre as instaladas."""
        alvo = IDIOMA_NARRACAO.lower()
        curto = alvo.split("-")[0]
        try:
            vozes = motor.getProperty("voices")
        except Exception:  # noqa: BLE001
            return None
        for voz in vozes:
            idiomas = [
                item.decode(errors="ignore") if isinstance(item, bytes) else str(item)
                for item in getattr(voz, "languages", []) or []
            ]
            texto = " ".join([voz.id, getattr(voz, "name", ""), *idiomas]).lower()
            if alvo in texto or f"_{curto}-" in texto or f" {curto} " in texto:
                return voz.id
        return None

File: web/ui/narrador.js
/**
 * Narração por voz do corpo focado, usando a Web Speech API do navegador.
 *
 * Equivalente web do ui/narrador.py — mesma frase, mesmo idioma, mesma regra de
 * substituir a fala anterior em vez de enfileirar. Aqui não há thread: a API já
 * é assíncrona, mas ela tem duas armadilhas próprias tratadas abaixo.
 */

import {
  ELEVENLABS_VOZ_NOME,
  ENDPOINT_VOZ,
  IDIOMA_NARRACAO,
  NARRACAO_ATIVA_PADRAO,
  NARRAR_FICHA_COMPLETA,
  URL_SERVIDOR_RENDER,
  VELOCIDADE_NARRACAO,
  VOLUME_NARRACAO,
} from "../config.js";
import { nomeDoTipo } from "./ficha.js";

const CHAVE_PREFERENCIA = "sistema-solar:narracao";
/** O pyttsx3 mede em palavras por minuto; a Web Speech usa 1 = natural. */
const PPM_NATURAL = 165;

// Artigo definido antes do nome. Em português corrente dizemos "o Sol" e "a
// Terra", mas "Marte" e "Júpiter" dispensam artigo.
const ARTIGO_DEFINIDO = { Sol: "O", Terra: "A", Lua: "A" };

// Descrição do tipo já com artigo indefinido, para a frase fechar concordância.
const TIPO_NARRADO = {
  estrela: "uma estrela",
  rochoso: "um planeta rochoso",
  gasoso: "um gigante gasoso",
  satelite: "um satélite natural",
};

/** Número no padrão pt-BR, que é como o sintetizador lê corretamente. */
function numero(valor, casas = 0) {
  return new Intl.NumberFormat("pt-BR", {
    minimumFractionDigits: casas,
    maximumFractionDigits: casas,
  }).format(valor);
}

/**
 * A ficha do corpo em orações completas, prontas para serem lidas.
 *
 * É a mesma informação do card, dita em português corrente: "Tem 12.756
 * quilômetros de diâmetro" em vez de "Diâmetro equatorial: 12.756 km". Rótulo e
 * valor soltos soariam como uma planilha sendo lida em voz alta.
 */
/**
 * Número para ser dito em voz alta, sem casas decimais inúteis.
 *
 * "243,0 dias" era lido como "duzentos e quarenta e três vírgula zero"; o zero
 * à direita não informa nada e atrapalha a fala.
 */
function quantidade(valor, casas = 1) {
  if (Math.abs(valor - Math.round(valor)) < 10 ** -casas / 2) {
    return numero(Math.round(valor));
  }
  return numero(valor, casas);
}

/** Escolhe singular ou plural conforme o número dito. */
function concordar(valor, singular, plural) {
  return Math.abs(valor - 1) < 1e-9 ? singular : plural;
}

/**
 * Distância em escala legível.
 *
 * Ninguém diz "149.600.000 quilômetros" em voz alta, e o sintetizador se perde
 * no meio dos nove dígitos — na transcrição saía "149.600 mil". Acima de um
 * milhão a frase passa a usar milhões e bilhões.
 */
function distanciaFalada(km) {
  if (km >= 1_000_000_000) {
    const bilhoes = km / 1_000_000_000;
    return `${quantidade(bilhoes, 2)} ${concordar(bilhoes, "bilhão", "bilhões")} de quilômetros`;
  }
  if (km >= 1_000_000) {
    const milhoes = km / 1_000_000;
    return `${quantidade(milhoes, 1)} ${concordar(milhoes, "milhão", "milhões")} de quilômetros`;
  }
  return `${numero(km)} quilômetros`;
}

/**
 * Temperatura com o sinal dito por extenso.
 * O "−" antes do número desaparecia na fala: −65 virava "65 graus".
 */
function temperaturaFalada(graus) {
  return graus < 0
    ? `${numero(Math.abs(graus))} graus Celsius negativos`
    : `${numero(graus)} graus Celsius`;
}

/** Frase das luas, com o gênero certo — "2 luas" saía como "dois luas". */
function luasFaladas(luas) {
  if (luas <= 0) return null;
  if (luas === 1) return "Tem uma lua conhecida.";
  if (luas === 2) return "Tem duas luas conhecidas.";
  return `Tem ${numero(luas)} luas conhecidas.`;
}

export function frasesDaFicha(corpo) {
  const frases = [`Tem ${numero(corpo.diametroKm)} quilômetros de diâmetro.`];

  if (corpo.orbitaEmTornoDe) {
    const preposicao = corpo.orbitaEmTornoDe === "Terra" ? "da" : "de";
    frases.push(
      `Fica a ${distanciaFalada(corpo.distanciaKm)} ${preposicao} ${corpo.orbitaEmTornoDe}.`,
    );
  } else if (corpo.distanciaUa > 0) {
    const ua = corpo.distanciaUa;
    frases.push(
      `Fica a ${quantidade(ua, 2)} ` +
        `${concordar(ua, "unidade astronômica", "unidades astronômicas")} do Sol, ` +
        `ou seja, ${distanciaFalada(corpo.distanciaKm)}.`,
    );
  }

  if (corpo.periodoOrbitalDias > 0) {
    if (corpo.periodoOrbitalDias >= 365.26) {
      const anos = corpo.periodoOrbitalDias / 365.26;
      frases.push(
        `Uma volta completa leva ${quantidade(anos, 1)} ` +
          `${concordar(anos, "ano terrestre", "anos terrestres")}.`,
      );
    } else {
      const dias = corpo.periodoOrbitalDias;
      frases.push(
        `Uma volta completa leva ${quantidade(dias, 0)} ${concordar(dias, "dia", "dias")}.`,
      );
    }
  }

  const horas = Math.abs(corpo.periodoRotacaoHoras);
  const sentido =
    corpo.periodoRotacaoHoras < 0 ? " no sentido contrário ao dos demais" : "";
  if (horas >= 48) {
    const diasRotacao = horas / 24;
    frases.push(
      `Gira em torno de si mesmo em ${quantidade(diasRotacao, 1)} ` +
        `${concordar(diasRotacao, "dia", "dias")}${sentido}.`,
    );
  } else {
    frases.push(
      `Gira em torno de si mesmo em ${quantidade(horas, 1)} ` +
        `${concordar(horas, "hora", "horas")}${sentido}.`,
    );
  }

  const luas = luasFaladas(corpo.luas);
  if (luas) frases.push(luas);

  frases.push(`A temperatura média é de ${temperaturaFalada(corpo.temperaturaMediaC)}.`);
  frases.push(corpo.fatoCurioso);
  return frases;
}

/**
 * Texto narrado ao focar um corpo.
 *
 * A abertura é uma **oração completa** ("A Terra é um planeta rochoso.") e não
 * uma lista de termos ("Terra. Planeta rochoso"). O motivo é prático: o modelo
 * de voz identifica o idioma pelo texto, e nomes latinos soltos são ambíguos —
 * medindo com a transcrição da própria ElevenLabs, "Sol. Estrela" saía em
 * espanhol e "Marte. Planeta rochoso" em inglês. O verbo "é" acentuado e os
 * artigos são o que ancora a frase no português.
 */
export function textoDoCorpo(corpo) {
  const artigo = ARTIGO_DEFINIDO[corpo.nome];
  const sujeito = artigo ? `${artigo} ${corpo.nome}` : corpo.nome;
  const descricao = TIPO_NARRADO[corpo.tipo] ?? "um corpo celeste";
  const partes = [`${sujeito} é ${descricao}.`];
  if (NARRAR_FICHA_COMPLETA) partes.push(...frasesDaFicha(corpo));
  return partes.join(" ");
}

export class Narrador {
  constructor() {
    this.disponivel = typeof window !== "undefined" && "speechSynthesis" in window;
    // A voz neural é tentada até dar errado uma vez; a do navegador é o piso.
    this._neuralDisponivel = typeof fetch === "function";
    this.ativo = (this.disponivel || this._neuralDisponivel) && this._preferenciaSalva();
    this._voz = null;
    this._audio = null;
    // Contador de pedidos: descarta o áudio que chega depois de o usuário já
    // ter escolhido outro planeta.
    this._pedidoAtual = 0;
    if (this.disponivel) this._carregarVozes();
  }

  /** Nome do backend em uso, para o HUD mostrar. */
  get backend() {
    return this._neuralDisponivel ? ELEVENLABS_VOZ_NOME : "voz do navegador";
  }

  /**
   * A lista de vozes costuma vir vazia na primeira chamada: o navegador a
   * carrega de forma assíncrona e avisa por `voiceschanged`.
   */
  _carregarVozes() {
    const escolher = () => {
      const vozes = window.speechSynthesis.getVoices();
      if (!vozes.length) return;
      const alvo = IDIOMA_NARRACAO.toLowerCase();
      const curto = alvo.split("-")[0];
      this._voz =
        vozes.find((v) => v.lang?.toLowerCase() === alvo) ??
        vozes.find((v) => v.lang?.toLowerCase().startsWith(curto)) ??
        null;
    };
    escolher();
    window.speechSynthesis.addEventListener("voiceschanged", escolher);
  }

  _preferenciaSalva() {
    try {
      const salvo = localStorage.getItem(CHAVE_PREFERENCIA);
      return salvo === null ? NARRACAO_ATIVA_PADRAO : salvo === "1";
    } catch {
      // localStorage bloqueado (modo privado, cookies desativados).
      return NARRACAO_ATIVA_PADRAO;
    }
  }

  _salvarPreferencia() {
    try {
      localStorage.setItem(CHAVE_PREFERENCIA, this.ativo ? "1" : "0");
    } catch {
      // Sem persistência a preferência vale só para esta sessão.
    }
  }

  /** Liga/desliga a narração e devolve o novo estado. */
  alternar() {
    if (!this.disponivel && !this._neuralDisponivel) return false;
    this.ativo = !this.ativo;
    if (!this.ativo) this._calar();
    this._salvarPreferencia();
    return this.ativo;
  }

  /**
   * Fala uma frase, cortando a anterior.
   *
   * Tenta primeiro a voz neural (ElevenLabs, via /api/voz) e cai para a voz do
   * navegador se a função não existir, não estiver configurada ou falhar.
   */
  anunciar(texto) {
    if (!this.ativo || !texto) return;
    this._calar();
    if (this._neuralDisponivel) {
      this._falarNeural(texto);
      return;
    }
    this._falarLocal(texto);
  }

  /**
   * Toca o áudio vindo de /api/voz.
   *
   * A rota devolve 503 quando o servidor não tem chave configurada — é o sinal
   * combinado para desistir da voz neural pelo resto da sessão e não repetir
   * uma ida à rede a cada troca de planeta.
   */
  async _falarNeural(texto) {
    const pedido = ++this._pedidoAtual;
    try {
      let resposta = await fetch(`${ENDPOINT_VOZ}?texto=${encodeURIComponent(texto)}`);
      if (!resposta.ok && URL_SERVIDOR_RENDER) {
        try {
          const respRender = await fetch(`${URL_SERVIDOR_RENDER}/api/voz?texto=${encodeURIComponent(texto)}`);
          if (respRender.ok) resposta = respRender;
        } catch {
          // ignora falha de rede do render
        }
      }
      if (!resposta.ok) {
        if (resposta.status === 503 || resposta.status === 404) {
          this._neuralDisponivel = false;
        }
        throw new Error(`HTTP ${resposta.status}`);
      }
      const blob = await resposta.blob();
      // Outro planeta foi escolhido enquanto o áudio baixava: este já não vale.
      if (pedido !== this._pedidoAtual || !this.ativo) return;
      const audio = new Audio(URL.createObjectURL(blob));
      audio.volume = VOLUME_NARRACAO;
      audio.addEventListener("ended", () => URL.revokeObjectURL(audio.src));
      this._audio = audio;
      await audio.play();
    } catch {
      if (pedido === this._pedidoAtual && this.ativo) this._falarLocal(texto);
    }
  }

  /** Voz do próprio navegador (Web Speech API). */
  _falarLocal(texto) {
    if (!this.disponivel) return;
    const fala = new SpeechSynthesisUtterance(texto);
    fala.lang = IDIOMA_NARRACAO;
    fala.rate = VELOCIDADE_NARRACAO / PPM_NATURAL;
    fala.volume = VOLUME_NARRACAO;
    if (this._voz) fala.voice = this._voz;
    window.speechSynthesis.speak(fala);
  }

  /** Interrompe qualquer fala em andamento, dos dois backends. */
  _calar() {
    this._pedidoAtual += 1;
    // Sem o cancel as falas empilham: trocando de planeta rápido o usuário
    // ouviria uma fila inteira de nomes antigos.
    if (this.disponivel) window.speechSynthesis.cancel();
    if (this._audio) {
      this._audio.pause();
      this._audio = null;
    }
  }

  /** Silencia imediatamente (usado ao sair da página). */
  parar() {
    this._calar();
  }
}
